src/actions/google/analytics/data_import.ts

```diff
@@ -0,0 +1,145 @@
+  readonly supportedFormattings = [Hub.ActionFormatting.Unformatted]
+  readonly supportedVisualizationFormattings = [Hub.ActionVisualizationFormatting.Noapply]
+  readonly supportedDownloadSettings = [Hub.ActionDownloadSettings.Url]
+  readonly usesStreaming = true
+  readonly requiredFields = []
+  readonly params = []
+
+  /******** OAuth properties ********/
+
+  readonly redirectUri = `${process.env.ACTION_HUB_BASE_URL}/actions/${encodeURIComponent(this.name)}/oauth_redirect`
+  readonly oauthClientId: string
+  readonly oauthClientSecret: string
+  readonly oauthScopes = ["https://www.googleapis.com/auth/analytics.edit"]
+  readonly oauthHelper: GoogleOAuthHelper
+
+  /******** Constructor & some helpers ********/
+
+  constructor(oauthClientId: string, oauthClientSecret: string) {
+    super()
+    this.oauthClientId = oauthClientId
+    this.oauthClientSecret = oauthClientSecret
+    this.oauthHelper = new GoogleOAuthHelper(this)
+  }
+
+  log(level: string, ...rest: any[]) {
+    return winston.log(level, LOG_PREFIX, ...rest)
+  }
+
+  makeOAuthClient(redirect?: string) {
+    redirect = redirect ? redirect : this.redirectUri
+    return this.oauthHelper.makeOAuthClient(redirect)
+  }
+
+  /******** Endpoints for Hub.OAuthAction ********/
+
+  async oauthUrl(redirectUri: string, encryptedState: string) {
+    return this.oauthHelper.oauthUrl(redirectUri, encryptedState)
+  }
+
+  async oauthFetchInfo(urlParams: { [key: string]: string }, redirectUri: string) {
+    return this.oauthHelper.oauthFetchInfo(urlParams, redirectUri)
+  }
+
+  async oauthCheck(_request: Hub.ActionRequest) {
+    // This part of Hub.OAuthAction is deprecated and unused
+    return true
+  }
+
+  /******** Main Action Endpoints ********/
+
+  async execute(hubReq: Hub.ActionRequest) {
+    const wrappedResp = new WrappedResponse(Hub.ActionResponse)
+
+    let currentStep = "action setup"
+    try {
+      // The worker constructor will do a bunch of validation for us
+      const gaWorker = await GoogleAnalyticsActionWorker.fromHubRequest(hubReq, this)
+
+      currentStep = "Data upload step"
+      await gaWorker.uploadData()
+      this.log("info", `${currentStep} completed.`)
+      this.log("debug", "New upload id=", gaWorker.newUploadId)
+
+      // Since the upload was successful, update the lastUsedFormParams in user state
+      gaWorker.setLastUsedFormParams()
+      wrappedResp.setUserState(gaWorker.userState)
+
+      if (gaWorker.isDeleteOtherFiles) {
+        currentStep = "Delete other files step"
+        await gaWorker.deleteOtherFiles()
+        this.log("info", `${currentStep} completed.`)
+      }
+
+      // All is well if we made it this far
+      this.log("info", "Execution completed successfully.")
+      return wrappedResp.returnSuccess()
+    } catch (err) {
+      this.log("error", "Execution error:", err.toString())
+      this.log("error", "Exeuction errror JSON:", JSON.stringify(err))
```

Comment:
I need to check whether `JSON.stringify(err)` will inadvertently log the headers associated with the request, which would include the oauth access token. I originally added this because the google api error objects contain a bunch of additional properties with more helpful info than just the message (e.g. on a 400 Bad Request, it will include a list of the fields with problems).